[FLINK-12556][e2e] Extend some end-to-end tests to run with custom (input) File System implementation

1u0 opened the pull request:
## What is the purpose of the change

This pull request extends some end-to-end tests to be independent of origin of input source (be it a local file, s3 or an artificial file system plugin), provided that they all serve the same test data content.

In particular, a new (dummy) file system plugin is introduced to check that different Flink cluster setups are working with additional (external) plugin in `lib` folder.

Later, those tests can be modified to check that new plugin system also works in different cluster setups ([FLINK-11952](https://issues.apache.org/jira/browse/FLINK-11952))

## Brief change log

  - Yarn kerberos docker test is simplified (some code reuse) without changing behavior;
  - batch wordcount test is generalized and now hosts (deleted) shaded hadoop and shaded presto tests;
  - introduced a dummy FS plugin that serves hardcoded read-only contents from memory;
  - add some test runs that use dummy FS plugin as input source.


## Verifying this change

This change added tests and can be verified as follows:

  - The modified end-to-end tests should pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFs.java:
```diff
@@ -0,0 +1,166 @@
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+class DummyFs extends FileSystem {
```
Rename to `DummyFileSystem`? One thing is that I generally do not like abbreviations another is that the pre-existing convention is to not abbreviate this.

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFs.java:
```diff
@@ -0,0 +1,166 @@
+
+	@Override
+	public boolean delete(final Path path, final boolean recursive) throws IOException {
+		checkNotNull(path, "path is null");
```
nit: I would drop `checkNotNull` for methods that `throw new UnsupportedOperationException();`

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFs.java:
```diff
@@ -0,0 +1,166 @@
+
+	@Override
+	public boolean isDistributedFS() {
+		return false;
```
shouldn't this be `return true`? All of the processes will see the same dummy files, right?

1u0 replied:
I haven't dug into semantics of this flag and how it is interpreted internally. The original implementation is heavily based on `LocalFileSystem`. But I have changed it to `true`, the batch wordcount test still pass.

pnowojski replied:
I guess it should be fine either way, but to me it looks like `true` would be more accurate and potentially less confusing for future developers

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFs.java:
```diff
@@ -0,0 +1,166 @@
+	@Nullable
+	private byte[] getDataByPath(Path path) {
+		URI uri = path.toUri();
+		String resourcePath = uri.getPath().substring(1);
```
why do we need `substring(1)`?

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFsFactory.java:
```diff
@@ -0,0 +1,52 @@
+ * A FileSystem implementation for integration testing purposes. Supports and serves read-only content from static
+ * key value map.
+ */
+public class DummyFsFactory implements FileSystemFactory {
```
ditto `Fs` -> `FileSystem`.

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFs.java:
```diff
@@ -0,0 +1,166 @@
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
```
Should we add a java doc with a reference to `DummyFileSystemFactory`?

pnowojski replied:
There is no clear convention. I follow up a rule, that I place a comment if something is not obvious based on the class name/line of code. Here `DummyFileSystem` doesn't explain in what why is it "dummy", so it needs some explanation. The problem is that a developer can encounter this `DummyFileSystem` either through `FileSystem` or `FileSystemFactory`, so I would place an explanation in one of them, and reference it in the other.

It hardly matters if class is a test class, or not.

1u0 replied:
There is a description in the factory class.

pnowojski replied:
Yes, but:
> The problem is that a developer can encounter this DummyFileSystem either through FileSystem or FileSystemFactory, so I would place an explanation in one of them, and reference it in the other.

I mean it's not obvious that if you wonder what `DummyFileSystem` is doing, that you should check a comment in `DummyFileSystemFactory` class.

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFsFileStatus.java:
```diff
@@ -0,0 +1,67 @@
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.Path;
+
+class DummyFsFileStatus implements FileStatus {
```
`DummyFileStatus`?  (pre-existing convention)

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFsInputStream.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.io.IOException;
+
+class DummyFsInputStream extends FSDataInputStream {
```
`DummyDataInputStream`? (pre-existing convention)

pnowojski commented on flink-end-to-end-tests/run-nightly-tests.sh:
```diff
@@ -88,7 +88,8 @@ run_test "Resuming Externalized Checkpoint after terminal failure (rocks, increm
 ################################################################################
 
-run_test "Running Kerberized YARN on Docker test " "$END_TO_END_DIR/test-scripts/test_yarn_kerberos_docker.sh"
+run_test "Running Kerberized YARN on Docker test (file input)" "$END_TO_END_DIR/test-scripts/test_yarn_kerberos_docker.sh file"
```
I think we could just replace `file` to `dummy-fs` in yarn docker test

pnowojski commented on flink-end-to-end-tests/run-pre-commit-tests.sh:
```diff
@@ -55,6 +55,7 @@ run_test "Streaming Python Wordcount end-to-end test" "$END_TO_END_DIR/test-scri
 run_test "Wordcount end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wordcount.sh file"
 run_test "Shaded Hadoop S3A end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wordcount.sh hadoop"
 run_test "Shaded Presto S3 end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wordcount.sh presto"
+run_test "Custom FS plugin end-to-end test" "$END_TO_END_DIR/test-scripts/test_batch_wordcount.sh dummy-fs"
```
We do not need to use `dummy-fs` here. To test plugins we can just existing `presto` and `hadoop` s3 variants.

pnowojski replied:
Hmm, testing a test class?

pnowojski commented on tools/travis/splits/split_container.sh:
```diff
@@ -43,8 +43,9 @@ echo "Flink distribution directory: $FLINK_DIR"
 
-run_test "Wordcount on Docker test" "$END_TO_END_DIR/test-scripts/test_docker_embedded_job.sh"
-run_test "Running Kerberized YARN on Docker test " "$END_TO_END_DIR/test-scripts/test_yarn_kerberos_docker.sh"
+run_test "Wordcount on Docker test (custom fs plugin)" "$END_TO_END_DIR/test-scripts/test_docker_embedded_job.sh dummy-fs"
```
ditto: I think we do not need to increase number of tests here

pnowojski commented on flink-end-to-end-tests/test-scripts/test_yarn_kerberos_docker.sh:
```diff
@@ -47,6 +65,10 @@ function cluster_shutdown {
 trap cluster_shutdown INT
 trap cluster_shutdown EXIT
 
+#if [[ 1 ]]; then
```
some leftovers?

1u0 commented on flink-end-to-end-tests/test-scripts/test_yarn_kerberos_docker.sh:
```diff
@@ -92,34 +88,28 @@ do
 fi
 
+mkdir -p $FLINK_TARBALL_DIR
+tar czf $FLINK_TARBALL_DIR/$FLINK_TARBALL -C $(dirname $FLINK_DIR) .
```
They are moved here to be run later. So if the test fails early (for example on cluster spin up), it won't create unused tarball.
This change was part of the test refactoring (you can check it by reading commits one-by-one).

pnowojski replied:
Yes, I was reading commits one by one and I was still confused (hence my question). I wasn't (and I'm still not) sure if all of those refactorings in the first commit are related to one another or if they are 3 independent things - if the latter one, it would be better to have them in separate commits.

pnowojski replied:
> I wasn't (and I'm still not) sure if all of those refactorings in the first commit are related to one another or if they are 3 independent things 

Bumping the question :)

1u0 commented on flink-end-to-end-tests/test-scripts/test_yarn_kerberos_docker.sh:
```diff
@@ -92,34 +88,28 @@ do
-    fi
-
-    echo "ERROR: Could not start hadoop cluster. Retrying..."
-    docker-compose -f $END_TO_END_DIR/test-scripts/docker-hadoop-secure-cluster/docker-compose.yml down
```
I expect it to be handled by `cluster_shutdown`.

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFs.java:
```diff
@@ -0,0 +1,166 @@
+
+	DummyFs(Map<String, String> contents) {
+		this.workingDir = new File(System.getProperty("user.dir")).toURI();
+		this.homeDir = new File(System.getProperty("user.home")).toURI();
```
It looks like it should be guaranteed to be set. At least I couldn't google out anything that would contradict this.

pnowojski commented:
@1u0 30 seconds is a lot for a single test :(

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFileSystemFileStatus.java:
```diff
@@ -0,0 +1,67 @@
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.Path;
+
+class DummyFileSystemFileStatus implements FileStatus {
```
`DummyFileStatus` would follow a pre-existing convention of `HadoopFileStatus` and `LocalFileStatus`

1u0 replied:
If you insist in such change, I propose to revert back the renaming and change `DummyFileSystem` to `DummyFs` with keeping it as prefix in those classes:
 1) I'm personally neutral about using common abbreviations (id, cpu, url, db, fs and so on). And there are already implementations of `FileSystem` that use such convention;
 2) the prefix here is intended rather as a marker prefix, similar to "Hadoop", "FlinkS3", "MapR" in other `FileSystem` implementations. It doesn't mean that the corresponding `FileStatus/FSDataInputStream` implementations are dummy or mocks.

pnowojski replied:
Ok I get your point. In that case I would vote for one of the following:
1. `DummyFileSystem`, `DumyFileStatus`, `DummyDataInputStream`
2. `DummyFSFileSystem` `DumyFSFileStatus`, `DummyFSDataInputStream` (note `FS` which would also feet the code base better compared to `Fs` and note that all files should have the same prefix)
3. Rename it altogether, for example to `StaticContent` or `PredefinedContent` (`StaticContentFileSystem`, `StaticContentFileStatus`, `StaticContentDataInputStream`, ...)

I think the 3rd would be the best, as it also describes in what way the file system is "dummy", not just the fact that it is "dummy".

1u0 replied:
I'm personally against option 1. As it's misleading that `DummyFileStatus`, `DummyDataInputStream`, ... are dummy, but actually they are valid implementations for corresponding FS.
I can adjust PR for option 2., by renaming `DummyFs` to `DummyFsFileSystem`. I was not aware about "abbreviation in uppercase" convention. In particular, I was using as examples of naming `HadoopFsFactory`, `MapRFsFactory`.

Initially, I also though about option 3, but `static` can't be used as a standalone part of package name. While `staticcontent` package name feels off style with other FSs.

pnowojski replied:
Ok, you are right about `HadoopFsFactory` and `MapRFsFactory`. I was looking mostly at different abbreviations. `DummyFsFileSystem` sounds good to me in that case.

pnowojski commented on flink-end-to-end-tests/flink-plugins-test/src/main/java/org/apache/flink/fs/dummy/DummyFileSystemInputStream.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.io.IOException;
+
+class DummyFileSystemInputStream extends FSDataInputStream {
```
ditto: `DummyInputStream`

1u0 commented:
Timings of the tests, from the CI logs:
```
[PASS] 'Running Kerberized YARN on Docker test (default input)' passed after 8 minutes and 13 seconds! Test exited with exit code 0.
[PASS] 'Running Kerberized YARN on Docker test (custom fs plugin)' passed after 4 minutes and 33 seconds! Test exited with exit code 0.

[PASS] 'Wordcount end-to-end test' passed after 0 minutes and 12 seconds! Test exited with exit code 0.
[PASS] 'Shaded Hadoop S3A end-to-end test' passed after 0 minutes and 1 seconds! Test exited with exit code 0.
[PASS] 'Shaded Presto S3 end-to-end test' passed after 0 minutes and 0 seconds! Test exited with exit code 0.
[PASS] 'Custom FS plugin end-to-end test' passed after 0 minutes and 13 seconds! Test exited with exit code 0.
```

So adding a modified Yarn kerberos test as addition, extends tests run by 4,5 minutes. He it's may look faster, because it reuses the docker build created by previous test run. Otherwise, I expect them to run about the same amount of time in equal conditions.

CC @zentol.

pnowojski commented:
LGTM % renaming, one question and if Chesney is fine with extending the testing time (I would keep the original count of tests)

1u0 commented:
@pnowojski, thanks for the review again. I have updated the PR and hopefully addressed your comments (except about additional test run time overhead).

If there is strong no, the latter can be also adjusted: both (or either of them), the word count with dummy-fs test (running on PR check) and the old yarn kerberos test with internal input (running nightly) can be removed from test runs.
I'm keeping them included for now, to at least showcase that both (old and new) tests are passed.

1u0 commented:
@pnowojski, I've rebased the PR.
The CI pass is green 🍀😃.
